### PR TITLE
Fix some warnings from clippy pedantic lints

### PR DESCRIPTION
**Pull request by duguorong009**
**What's been done:**
Fix some warnings from rust clippy pedantic lints(`cargo clippy -- -W clippy::pedantic`).

**Reference:** #513

### CODE REVIEW

File: circuit-construction/src/writer.rs [hunk 1]
```diff
@@ -782,7 +790,7 @@ pub trait Cs<F: PrimeField> {
                                         // Do one full round on the previous value
                                         let mut acc = states[states.len() - 1]
                                             .iter()
-                                            .map(|x| x.val())
+                                            .map(crate::writer::Var::val)
```
Comment (mimoo) [inline circuit-construction/src/writer.rs:793]: don't think this is a good change :o

Comment (duguorong009) [reply circuit-construction/src/writer.rs:793]: Do you think that way?
So, you mean better to keep original style? like `(|x| x.val())`

File: kimchi/src/verifier.rs [hunk 2]
```diff
@@ -616,7 +627,11 @@ where
                         panic!("runtime lookup table is unused in the linearization")
                     }
                     Index(t) => {
-                        use GateType::*;
+                        use GateType::{
+                            CairoClaim, CairoFlags, CairoInstruction, CairoTransition, ChaCha0,
+                            ChaCha1, ChaCha2, ChaChaFinal, CompleteAdd, EndoMul, EndoMulScalar,
+                            Generic, Lookup, Poseidon, RangeCheck0, RangeCheck1, VarBaseMul, Zero,
+                        };
```
Comment (mimoo) [inline kimchi/src/verifier.rs:634]: not a fan of that change also, I find that in general it is fine to open with a wildcard if it's an enum and right before using the enum

**Comment by mimoo** [comment]
This is great!

So, going through this list I have several reactions:

1. a lot of these fixes are useful (e.g. documenting panics, removing `_` from match statements, removing wildcard imports, etc.)
2. some of these seem less useful and I'm wondering if we should keep them

I guess 2. prompts me to ask: should we enforce pedantic clippy lints in CI? I'm thinking that it's probably too much, and a better approach would be to selectively enable _some_ lints manually in CI. (Back in my previous job, we would review new clippy lints from times to times and figure out which one would be useful to run, it was neat!)

If we don't enable pedantic lints in CI, then a number of lints here are debatable. Here's some thoughts:

* are the `#[must_use]` annotations really useful? CI already enforce a clean compiler output, and the compiler will warn you if you don't use a return value already
* the new `#[allow(clippy::_)]` would not be necessary
* I find importing enums with a wildcard OK if it's right before using it personally, I find it pretty verbose to exhaustively import enum variants

**Comment by duguorong009** [comment]
First, thanks for the reviews.
It helps me to understand the code style you(or team) follow.
Also, personally, I learnt much of the codebase while checking & fixing the lint warnings, especially documentation. 😄 

1. Yes, I also think we don't need pedantic lints in CI. That's too much, and let developers use up much more time.
2. About your thoughts(in case of don't enable pedantic lints in CI):
   - I also think `#[must_use]` and `#[allow(clippy::_)] should be used in limit way.
     Atm, I just tried to fix the lint warnings as much as possible.
   - I also think wildcard is okay for importing several enum or struct from other packages.
     Maybe, it is better to use the wildcard if the number of imported enum/struct is bigger than 3, I think.

**Comment by duguorong009** [comment]
So, @mimoo , let me know your idea on my suggestions.
Simply, I accept your ideas, and added some more detail.
About the current pedantic lints fixture commits, do you think I should correct in the way we are talking, before merge?

**Comment by mimoo** [comment]
Ah yeah, I guess if you agree with me can you remove all the ones we said were not necessary and then I can rereview it to make sure it all makes sense?

**Comment by duguorong009** [comment]
okay, got it.
Will update the PR as discussed.

**Comment by duguorong009** [comment]
@mimoo 
Please go for re-review the PR.

File: kimchi/src/circuits/expr.rs [hunk 3]
```diff
@@ -2168,12 +2174,15 @@ pub mod constraints {
         Self: std::marker::Sized,
     {
         /// Double the value
+
```
Comment (mimoo) [inline kimchi/src/circuits/expr.rs:2177]: why the spaces there :o ?

File: kimchi/src/circuits/expr.rs [hunk 4]
```diff
@@ -2333,7 +2346,7 @@ pub fn coeff<F>(i: usize) -> E<F> {
 
 /// You can import this module like `use kimchi::circuits::expr::prologue::*` to obtain a number of handy aliases and helpers
 pub mod prologue {
-    pub use super::{coeff, constant, index, witness, witness_curr, witness_next, E};
+    pub use super::*;
```
Comment (mimoo) [inline kimchi/src/circuits/expr.rs:2349]: we actually want to be explicit here

File: kimchi/src/circuits/gate.rs [hunk 5]
```diff
@@ -43,6 +43,7 @@ impl CurrOrNext {
     /// Compute the offset corresponding to the `CurrOrNext` value.
     /// - `Curr.shift() == 0`
     /// - `Next.shift() == 1`
+
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:46]: I don't think it's idiomatic to put a space between a doc comment and a function :o

File: kimchi/src/circuits/gate.rs [hunk 6]
```diff
@@ -190,6 +191,7 @@ impl<F: PrimeField> ToBytes for CircuitGate<F> {
 
 impl<F: PrimeField> CircuitGate<F> {
     /// this function creates "empty" circuit gate
+
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:194]: here as well

File: kimchi/src/circuits/gate.rs [hunk 7]
```diff
@@ -200,14 +202,22 @@ impl<F: PrimeField> CircuitGate<F> {
...
         witness: &[Vec<F>; COLUMNS],
         cs: &ConstraintSystem<F>,
         public: &[F],
     ) -> Result<(), String> {
-        use GateType::*;
+        use GateType::{
+            CairoClaim, CairoFlags, CairoInstruction, CairoTransition, ChaCha0, ChaCha1, ChaCha2,
+            ChaChaFinal, CompleteAdd, EndoMul, EndoMulScalar, Generic, Lookup, Poseidon,
+            RangeCheck0, RangeCheck1, VarBaseMul, Zero,
+        };
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:220]: better with a wildcard

File: kimchi/src/circuits/gate.rs [hunk 8]
```diff
@@ -477,6 +487,8 @@ pub mod caml {
 
 #[cfg(test)]
 mod tests {
+    use crate::circuits::wires::PERMUTS;
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:490]: ?

File: kimchi/src/circuits/lookup/constraints.rs [hunk 9]
```diff
@@ -1,6 +1,9 @@
 use crate::{
     circuits::{
-        expr::{prologue::*, Column, ConstantExpr},
+        expr::{
+            prologue::{witness, E},
```
Comment (mimoo) [inline kimchi/src/circuits/lookup/constraints.rs:4]: btw prologue is supposed to be prelude here (I think I was mixing up the two words), and as such is supposed to be imported with a wildcard (that's what `prelude`s are for)

File: kimchi/src/circuits/lookup/runtime_tables.rs [hunk 10]
```diff
@@ -1,9 +1,9 @@
 //! Runtime tables are tables (or arrays) that can be produced during proof creation.
-//! The setup has to prepare for their presence using [RuntimeTableCfg].
-//! At proving time, the prover can use [RuntimeTable] to specify the actual tables.
+//! The setup has to prepare for their presence using [`RuntimeTableCfg`].
+//! At proving time, the prover can use [`RuntimeTable`] to specify the actual tables.
 
 use crate::circuits::{
-    expr::{prologue::*, Column},
+    expr::{prologue::E, Column},
```
Comment (mimoo) [inline kimchi/src/circuits/lookup/runtime_tables.rs:6]: same here

File: kimchi/src/prover.rs [hunk 11]
```diff
@@ -135,7 +139,20 @@ where
...
+    /// Will give error if inputs(like `lookup_context.joint_lookup_table_d8`) are None.
+    ///
+    /// # Panics
+    ///
+    /// Will panic if `lookup_context.joint_lookup_table_d8` is None.
+    #[allow(
+        clippy::needless_pass_by_value,
+        clippy::too_many_lines,
+        clippy::cast_possible_truncation
+    )]
```
Comment (mimoo) [inline kimchi/src/prover.rs:155]: there's still some clippy stuff :o

File: kimchi/src/tests/framework.rs [hunk 12]
```diff
@@ -52,37 +52,31 @@ pub(crate) struct TestFramework {
...
     }
 
-    #[must_use]
     pub(crate) fn witness(mut self, witness: [Vec<Fp>; COLUMNS]) -> Self {
         self.witness = Some(witness);
         self
     }
 
-    #[must_use]
     pub(crate) fn public_inputs(mut self, public_inputs: Vec<Fp>) -> Self {
```
Comment (mimoo) [inline kimchi/src/tests/framework.rs:65]: actually, now I realize that I don't think the compiler will yell at you if there isn't a `#[must_use]`, I think the compiler yells at you when you don't use a type `Result`. In this API I think it makes sense to leave the #[must_use]

File: tools/kimchi-visu/src/lib.rs [hunk 13]
```diff
@@ -44,7 +44,10 @@ where
     F: PrimeField,
 {
     fn latex() -> Vec<Vec<String>> {
-        Self::constraints().iter().map(|c| c.latex_str()).collect()
+        Self::constraints()
+            .iter()
+            .map(kimchi::circuits::expr::Expr::latex_str)
```
Comment (mimoo) [inline tools/kimchi-visu/src/lib.rs:49]: ^

Comment (duguorong009) [reply tools/kimchi-visu/src/lib.rs]: Done https://github.com/o1-labs/proof-systems/pull/752/commits/e74530b7cf67a9cf66787be09ff530dff090630e

File: kimchi/src/tests/framework.rs (hunk 12 above)
Comment (duguorong009) [reply kimchi/src/tests/framework.rs]: Done https://github.com/o1-labs/proof-systems/pull/752/commits/e74530b7cf67a9cf66787be09ff530dff090630e

File: kimchi/src/prover.rs (hunk 11 above)
Comment (duguorong009) [reply kimchi/src/prover.rs]: Done https://github.com/o1-labs/proof-systems/pull/752/commits/e74530b7cf67a9cf66787be09ff530dff090630e

File: kimchi/src/circuits/lookup/runtime_tables.rs (hunk 10 above)
Comment (duguorong009) [reply kimchi/src/circuits/lookup/runtime_tables.rs]: Done https://github.com/o1-labs/proof-systems/pull/752/commits/e74530b7cf67a9cf66787be09ff530dff090630e

File: kimchi/src/circuits/lookup/constraints.rs (hunk 9 above)
Comment (duguorong009) [reply kimchi/src/circuits/lookup/constraints.rs]: Done https://github.com/o1-labs/proof-systems/pull/752/commits/e74530b7cf67a9cf66787be09ff530dff090630e

File: kimchi/src/circuits/gate.rs (hunk 7 above)
Comment (duguorong009) [reply kimchi/src/circuits/gate.rs]: Done https://github.com/o1-labs/proof-systems/pull/752/commits/e74530b7cf67a9cf66787be09ff530dff090630e

File: kimchi/src/circuits/gate.rs (hunk 6 above)
Comment (duguorong009) [reply kimchi/src/circuits/gate.rs]: Done https://github.com/o1-labs/proof-systems/pull/752/commits/e74530b7cf67a9cf66787be09ff530dff090630e

File: kimchi/src/circuits/gate.rs (hunk 5 above)
Comment (duguorong009) [reply kimchi/src/circuits/gate.rs]: Done https://github.com/o1-labs/proof-systems/pull/752/commits/e74530b7cf67a9cf66787be09ff530dff090630e

File: kimchi/src/circuits/expr.rs (hunk 4 above)
Comment (duguorong009) [reply kimchi/src/circuits/expr.rs]: Done https://github.com/o1-labs/proof-systems/pull/752/commits/e74530b7cf67a9cf66787be09ff530dff090630e

File: kimchi/src/circuits/expr.rs (hunk 3 above)
Comment (duguorong009) [reply kimchi/src/circuits/expr.rs]: Done https://github.com/o1-labs/proof-systems/pull/752/commits/e74530b7cf67a9cf66787be09ff530dff090630e

File: kimchi/src/circuits/gate.rs (hunk 8 above)
Comment (duguorong009) [reply kimchi/src/circuits/gate.rs]: Done https://github.com/o1-labs/proof-systems/pull/752/commits/a3f5b22770f98cd7bb2c6a188814eab4371d723e

**Comment by duguorong009** [comment]
Thanks for the comments! @mimoo 
In fact, I pushed the quick PR with what I could do instantly.

Note: I have not removed all the clippy allow lints. I just left the original ones.
Simply, I just tried to remove the allow Iints I pushed.

**Comment by duguorong009** [comment]
Finally, updated the PR as best as possible. :sweat_smile: 
Plz go for re-review. @mimoo

**Review by mimoo — APPROVED** [review]
woop, that was a big PR, thanks a lot for fixing all of these things :)

**Pull request merged**